topi/python/topi/cuda/conv2d.py

```diff
@@ -1,24 +1,32 @@
-# pylint: disable=invalid-name, no-member, too-many-locals, too-many-statements, too-many-arguments, too-many-branches, line-too-long
+# pylint: disable=invalid-name
 """Compute definition for conv2d with cuda backend"""
 import tvm
+from tvm import autotvm
 from tvm.contrib import cudnn
-import topi
-from ..nn.conv2d import conv2d
-from ..util import get_const_int
 
-@conv2d.register("cuda")
-def conv2d_cuda(data, kernel, stride, padding, layout='NCHW', out_dtype='float32'):
+from .. import nn, generic
+from ..util import get_const_int, get_const_tuple, traverse_inline
+
+from .conv2d_direct import schedule_direct_cuda
+from .conv2d_winograd import winograd_cuda, schedule_winograd_cuda
+
+
+@autotvm.task.register_topi_compute(nn.conv2d, ['cuda', 'gpu'], ['direct', 'winograd'])
+def conv2d_cuda(cfg, data, kernel, strides, padding, layout='NCHW', out_dtype='float32'):
     """Conv2D operator for cuda backend.
 
     Parameters
     ----------
-    input : tvm.Tensor
+    cfg: ConfigEntity
+        The config for this template
+
+    data : tvm.Tensor
         4-D with shape [batch, in_channel, in_height, in_width]
 
```

Comment:
is batch size > 1 supported? If not, add assertion or warning.

Reply:
It supports. AutoTVM templates support naive fallback.
For this template, we do tiling on filter, height, width dimension. In fallback mode, the tile sizes are all 1. i.e. we always let one block has one thread, and one thread compute one output.

After this PR, all backends support arbitrary shapes. I will add the batch size > 1 test cases in `topi_test_conv2d_nchw`

Reply:
We will also check in another template soon with batch size > 1.